controllers/AuthController.js

```diff
@@ -1,46 +1,95 @@
-import users from '../utils/sample.users';
+import bcrypt from 'bcrypt';
+
+import Model from '../models/Model';
+import { InternalServerError, CriticalError } from '../utils/errorHandlers';
+import { dev_logger } from '../utils/loggers';
+
+// import { dev_logger, test_logger } from '../utils/loggers';
+
+const users_model = new Model('users');
 
 const AuthController = {
     signup: (req, res) => {
-            
-        if (req.method === 'GET') {
-            res.render('authentication', { title: 'Sign Up' });
-        }
-        else {
-            const { email, password } = req.body;
-
-            res.status(200).json({
-                status: 201,
-                data: {
-                    id: 2,
-                    email,
-                    password,
-                    first_name: '',
-                    last_name: '',
-                    address: {
-                        home: '',
-                        office: '',
-                    },
-                    status: 'unverified',
-                    isAdmin: false,
-                    token: req.token,
+        const { email, password } = req.body;
+        const hashedPassword = bcrypt.hashSync(password, 8);
+        dev_logger(`token: ${req.token}`);
+
+        // check if user exists
+        const check_user_existence = async (email) => {
+            try {
+                const { rows } = await users_model.select(
+                    'id, email', `email='${email}'`);
+                const [ user, ] = rows;
+                if (user) return user;
+            }
+            catch (e) { return InternalServerError(req, res, e);}
+        };
+
+        // add user to db
+        const create_user = async (email, password) => {
+            try {
+                await users_model.insert('(email, password)',
+                    `'${email}', '${password}'`);                    
+            }
+            catch (e) { return InternalServerError(req, res, e);}
+        };
+
+        const get_new_user = async (email) => {
```

Comment:
These are helper functions. You can move them into a helper folder or something. Let the controller handle the logic and not function declarations.

Reply:
Done